### PR TITLE
PRM-856: added inset text

### PR DESCRIPTION
**Pull request by BackRowOla**

### CODE REVIEW

File: app/controllers/OtherReasonController.scala [hunk 1]
```diff
@@ -137,11 +137,12 @@ class OtherReasonController @Inject()(whenDidBecomeUnablePage: WhenDidBecomeUnab
             val previousUploadsWithoutDownloadURL = previousUploads.map(_.copy(downloadUrl = None))
             Json.stringify(Json.toJson(previousUploadsWithoutDownloadURL))
           })
+          val containsDuplicate = previousUploads.contains("isDuplicateOf")
```
Comment (LewisMitchell) [inline app/controllers/OtherReasonController.scala:140]: Do you still need this?

File: app/controllers/OtherReasonController.scala [hunk 2]
```diff
@@ -137,11 +137,12 @@ class OtherReasonController @Inject()(whenDidBecomeUnablePage: WhenDidBecomeUnab
...
           val initiateNextUploadUrl = controllers.routes.UpscanController.initiateCallToUpscan(request.session.get(SessionKeys.journeyId).get)
           val getStatusUrl = controllers.routes.UpscanController.getStatusOfFileUpload(request.session.get(SessionKeys.journeyId).get, _)
           val removeFileUrl = controllers.routes.UpscanController.removeFile(request.session.get(SessionKeys.journeyId).get, _)
           val postAction = navigation.nextPage(EvidencePage, mode)
-          Ok(uploadEvidencePage(postAction, initiateNextUploadUrl, getStatusUrl, removeFileUrl, previousUploads))
+          Ok(uploadEvidencePage(postAction, initiateNextUploadUrl, getStatusUrl, removeFileUrl, previousUploads, containsDuplicate))
```
Comment (LewisMitchell) [inline app/controllers/OtherReasonController.scala:145]: Do you still need this?

File: app/controllers/internal/UpscanCallbackController.scala [hunk 3]
```diff
@@ -45,13 +45,10 @@ class UpscanCallbackController @Inject()(repository: UploadJourneyRepository)
               seqOfChecksums => {
                 if(seqOfChecksums.contains(callbackModel.uploadDetails.get.checksum)) {
                   logger.debug("[UpscanCallbackController][callbackFromUpscan] - Checksum is already in Mongo. Marking file as duplicate.")
-                  val duplicateCallbackModel = callbackModel.copy(
-                    fileStatus = UploadStatusEnum.FAILED,
-                    failureDetails = Some(FailureDetails(FailureReasonEnum.DUPLICATE, "upscan.duplicateFile")),
```
Comment (LewisMitchell) [inline app/controllers/internal/UpscanCallbackController.scala:50]: There is a case statement for getting the localised message, you may want to remove it now as it's not needed:
https://github.com/hmrc/penalties-appeals-frontend/blob/0fd60ff0937b953b9fa5331d8461fa14c75baecb/app/helpers/UpscanMessageHelper.scala#L29

File: app/repositories/UploadJourneyRepository.scala [hunk 4]
```diff
@@ -115,4 +115,31 @@ class UploadJourneyRepository @Inject()(
         ).collect { case Some(x) => x }
       ).getOrElse(Seq.empty))
   }
+
+  def getFileNameFromChecksum(journeyId: Option[String], checkSum: String): Future[Option[Seq[Option[String]]]] = {
```
Comment (LewisMitchell) [inline app/repositories/UploadJourneyRepository.scala:119]: Do you still need this?

File: app/repositories/UploadJourneyRepository.scala [hunk 5]
```diff
@@ -115,4 +115,31 @@ class UploadJourneyRepository @Inject()(
...
+            upload => {
+              if (checkSum.equals(upload.checksum)) upload.fileName else ""
+            }
+          )
+        )
+      )
+    )
+  }
+
+  def isTheChecksumAndFileNameThenSame(jounreyId: Option[String], fileName: String, checkSum: String): Future[Boolean] = {
```
Comment (LewisMitchell) [inline app/repositories/UploadJourneyRepository.scala:133]: Do you still need this?

File: app/views/reasonableExcuseJourneys/other/UploadEvidencePage.scala.html [hunk 6]
```diff
@@ -17,7 +17,8 @@
         initiateNextFileUpload: Call,
         getStatusUrl: String => Call,
         removeFileUrl: String => Call,
-        previousUploads: String)(implicit request: Request[_], messages: Messages, appConfig: AppConfig)
+        previousUploads: String,
+        containsDuplicate: Boolean)(implicit request: Request[_], messages: Messages, appConfig: AppConfig)
```
Comment (LewisMitchell) [inline app/views/reasonableExcuseJourneys/other/UploadEvidencePage.scala.html:21]: Do you still need this?

File: it/controllers/internal/UpscanCallbackControllerISpec.scala [hunk 7]
```diff
@@ -178,6 +212,26 @@ class UpscanCallbackControllerISpec extends IntegrationSpecCommonBase {
         duplicateModelInRepository.fileStatus shouldBe duplicateJsonAsModel.fileStatus
         duplicateModelInRepository.uploadDetails shouldBe duplicateJsonAsModel.uploadDetails
       }
+
+      "the body received is valid by the file has the same checksum as another file - duplicate different filename" in {
```
Comment (LewisMitchell) [inline it/controllers/internal/UpscanCallbackControllerISpec.scala:216]: Is this still needed?

File: it/controllers/internal/UpscanCallbackControllerISpec.scala [hunk 8]
```diff
@@ -92,13 +109,30 @@ class UpscanCallbackControllerISpec extends IntegrationSpecCommonBase {
...
+        fileName = "file1.txt",
+        fileMimeType = "text/plain",
+        uploadTimestamp = LocalDateTime.of(2018, 4, 24, 9, 31),
+        checksum = "check12345678",
+        size = 987
+      )
+    )
+  )
+
+  val duplicateCheckSumJsonAsModel: UploadJourney = UploadJourney(
```
Comment (LewisMitchell) [inline it/controllers/internal/UpscanCallbackControllerISpec.scala:125]: Is this still needed?

File: test/views/UploadEvidencePageSpec.scala [hunk 9]
```diff
@@ -76,7 +76,8 @@ class UploadEvidencePageSpec extends SpecBase with ViewBehaviours {
         controllers.routes.UpscanController.initiateCallToUpscan("1234"),
         controllers.routes.UpscanController.getStatusOfFileUpload("1234", _),
         controllers.routes.UpscanController.removeFile("1234", _),
-        previousUploadsToString(previousUploads))(request, implicitly, implicitly)
+        previousUploadsToString(previousUploads),
+        false)(request, implicitly, implicitly)
```
Comment (LewisMitchell) [inline test/views/UploadEvidencePageSpec.scala:80]: Is this still needed?

File: test/views/UploadListPageSpec.scala [hunk 10]
```diff
@@ -57,34 +74,68 @@ class UploadListPageSpec extends SpecBase with ViewBehaviours {
...
+
+      val expectedContent = Seq(
+        Selectors.title -> titleMaxFiles,
+        Selectors.h1 -> h1MaxFiles,
+        Selectors.insetText -> insetTextMsg("1", "2"),
+        Selectors.button -> continueButton
+      )
+
+       behave like pageWithExpectedMessages(expectedContent)
+    }
```
Comment (LewisMitchell) [inline test/views/UploadListPageSpec.scala:123]: Could we have a test for multiple duplicates uploaded of the same set i.e. 
Document 1 has the same contents as Document 2, 3 and 4. You can remove duplicate documents using the ‘Remove‘ link.

File: app/viewtils/EvidenceFileUploadsHelper.scala [hunk 11]
```diff
@@ -16,19 +16,48 @@
...
+class EvidenceFileUploadsHelper @Inject()(govukInsetText: GovukInsetText,
+                                          uploadList: uploadList) {
 
   def displayContentForFileUploads(uploadedFiles: Seq[(UploadJourney, Int)], mode: Mode)(implicit messages: Messages, request: UserRequest[_]): Seq[Html] = {
     uploadedFiles.map(uploadWithIndex =>
       uploadList(uploadWithIndex._1.reference, uploadWithIndex._1.uploadDetails.get.fileName, uploadWithIndex._2 + 1, mode)
     )
   }
+
+  def duplicateDocumentInset(uploadedFiles: Seq[(UploadJourney, Int)])(
```
Comment (LewisMitchell) [inline app/viewtils/EvidenceFileUploadsHelper.scala:38]: Could we have a test for this?

File: test/views/UploadListPageSpec.scala [hunk 12]
```diff
@@ -29,22 +29,39 @@ import viewtils.RadioOptionHelper
 
 class UploadListPageSpec extends SpecBase with ViewBehaviours {
```
Comment (LewisMitchell) [inline test/views/UploadListPageSpec.scala:30]: Could we also have a test for the inset text not appearing when there is no duplicates? Maybe this could be added to an existing test rather than creating a new one?

**Comment by LewisMitchell** [comment]
Currently fails AT's which is expected. Could you also rename the PR?

**Review by LewisMitchell — CHANGES_REQUESTED** [review]

File: app/controllers/OtherReasonController.scala (hunk 1 above)
Comment (BackRowOla) [reply app/controllers/OtherReasonController.scala:140]: Not needed, will clean up

File: app/controllers/OtherReasonController.scala (hunk 2 above)
Comment (BackRowOla) [reply app/controllers/OtherReasonController.scala:145]: same as above

File: it/controllers/internal/UpscanCallbackControllerISpec.scala (hunk 7 above)
Comment (BackRowOla) [reply it/controllers/internal/UpscanCallbackControllerISpec.scala:216]: Added this test to cover the error same checksum, different filename case, but regardless of if the filename is the same or not, it'll still be classed as a duplicate so I'll remove it.

File: app/controllers/OtherReasonController.scala [hunk 13]
```diff
@@ -242,9 +244,10 @@ class OtherReasonController @Inject()(whenDidBecomeUnablePage: WhenDidBecomeUnab
             previousUploadsState <- uploadJourneyRepository.getUploadsForJourney(request.session.get(SessionKeys.journeyId))
           } yield {
             val uploadedFileNames: Seq[UploadJourney] = previousUploadsState.fold[Seq[UploadJourney]](Seq.empty)(_.filter(_.fileStatus == UploadStatusEnum.READY))
```
Comment (LewisMitchell) [inline app/controllers/OtherReasonController.scala:246]: You will need to update this as well  similar to the above:
```          
val uploadedFileNames: Seq[UploadJourney] = previousUploadsState.fold[Seq[UploadJourney]](Seq.empty)(_.filter( upload =>
            upload.fileStatus == UploadStatusEnum.READY || upload.fileStatus == UploadStatusEnum.DUPLICATE))
val optDuplicateFiles: Option[Html] = evidenceFileUploadsHelper.duplicateDocumentInset(uploadedFileNames.zipWithIndex)```

**Review by LewisMitchell — CHANGES_REQUESTED** [review]

File: app/viewtils/EvidenceFileUploadsHelper.scala [hunk 14]
```diff
@@ -16,19 +16,48 @@
...
+    implicit messages: Messages, request: UserRequest[_]): Option[Html] = {
+
+    val mappedDuplicates = uploadedFiles.groupBy(_._1.uploadDetails.map(_.checksum))
+
+    if (!mappedDuplicates.exists(_._2.size > 1)) {
```
Comment (LewisMitchell) [inline app/viewtils/EvidenceFileUploadsHelper.scala:43]: I'm just thinking about when we have WAITING uploads? If we look at the JS version, we could have multiple uploads that are WAITING and we might get None -> Seq(..., ...) which could cause a logical error - maybe if we filter the uploads?

File: app/viewtils/EvidenceFileUploadsHelper.scala [hunk 15]
```diff
@@ -16,19 +16,48 @@
...
+      None
+    } else if (mappedDuplicates.count(_._2.size > 1) > 1) {
+      val noOfDuplicateDocs = mappedDuplicates.filter(_._2.size > 1).map(_._2.size).sum
+      Some(govukInsetText(InsetText(content = Text(messages("otherReason.uploadList.multipleDuplicateInsetText", noOfDuplicateDocs)))))
+    } else {
+      val duplicates: Seq[(UploadJourney, Int)] = mappedDuplicates.filter(_._2.size > 1).head._2
+      val firstFile: Int = duplicates.head._2 + 1
+      val otherUploads = duplicates.tail
+      val allOtherUploads = otherUploads.map(_._2 + 1)
+      val insetTextMsg = allOtherUploads.size match {
```
Comment (LewisMitchell) [inline app/viewtils/EvidenceFileUploadsHelper.scala:53]: Just thinking if we could add some types in this method?

**Review by LewisMitchell — CHANGES_REQUESTED** [review]

File: app/viewtils/EvidenceFileUploadsHelper.scala (hunk 14 above)
Comment (BackRowOla) [reply app/viewtils/EvidenceFileUploadsHelper.scala:43]: The uploads are already being filtered here 
```
          val uploadedFileNames: Seq[UploadJourney] = previousUploadsState.fold[Seq[UploadJourney]](Seq.empty)(_.filter( upload =>
            upload.fileStatus == UploadStatusEnum.READY || upload.fileStatus == UploadStatusEnum.DUPLICATE))
```
in the controller before being passed into the to the helper function. So I don't think any marked as waiting would get through to this part of the function. Unless I'm misunderstanding things 😅

Comment (LewisMitchell) [reply app/viewtils/EvidenceFileUploadsHelper.scala:43]: Doh! Yes you are right! It was Friday 😉

File: app/viewtils/EvidenceFileUploadsHelper.scala (hunk 15 above)
Comment (BackRowOla) [reply app/viewtils/EvidenceFileUploadsHelper.scala:53]: Just to clarify, do you mean adding types to the cases in the match statement or the val declarations?

Comment (LewisMitchell) [reply app/viewtils/EvidenceFileUploadsHelper.scala:53]: Just in the val's will do 😄

**Comment by LewisMitchell** [comment]
Will need rebasing 😄

**Comment by BackRowOla** [comment]
> Will need rebasing 😄

On it 👍

File: app/controllers/OtherReasonController.scala [hunk 16]
```diff
@@ -43,6 +43,7 @@ import views.html.reasonableExcuseJourneys.other._
 import views.html.reasonableExcuseJourneys.other.noJs.{UploadAnotherDocumentPage, UploadFirstDocumentPage, UploadListPage, UploadTakingLongerThanExpectedPage}
 import viewtils.{EvidenceFileUploadsHelper, RadioOptionHelper}
 
+import java.awt.Color
```
Comment (LewisMitchell) [inline app/controllers/OtherReasonController.scala:46]: Could you remove this import

File: app/controllers/internal/UpscanCallbackController.scala [hunk 17]
```diff
@@ -45,10 +45,10 @@ class UpscanCallbackController @Inject()(repository: UploadJourneyRepository)
               seqOfChecksums => {
                 if(seqOfChecksums.contains(callbackModel.uploadDetails.get.checksum)) {
                   logger.debug("[UpscanCallbackController][callbackFromUpscan] - Checksum is already in Mongo. Marking file as duplicate.")
-                  val duplicateCallbackModel = callbackModel.copy(
-                    fileStatus = UploadStatusEnum.DUPLICATE
-                  )
-                  repository.updateStateOfFileUpload(journeyId, duplicateCallbackModel).map(_ => NoContent)
+                      val duplicateCallbackModel = callbackModel.copy(
```
Comment (LewisMitchell) [inline app/controllers/internal/UpscanCallbackController.scala:48]: Could you unindent this?

File: it/controllers/internal/UpscanCallbackControllerISpec.scala (hunk 8 above)
Comment (LewisMitchell) [reply it/controllers/internal/UpscanCallbackControllerISpec.scala:124]: Is this still needed?

File: it/controllers/internal/UpscanCallbackControllerISpec.scala [hunk 18]
```diff
@@ -64,6 +64,23 @@ class UpscanCallbackControllerISpec extends IntegrationSpecCommonBase {
       |""".stripMargin
   )
 
+  val duplicateFilCheckSumAsJson: JsValue = Json.parse(
```
Comment (LewisMitchell) [inline it/controllers/internal/UpscanCallbackControllerISpec.scala:67]: Is this still needed?

**Review by LewisMitchell — CHANGES_REQUESTED** [review]
I feel like there is a test or 2 missing in the controller integration test? Could we add some tests that shows the inset text?
So like:
2 duplicates (of the same file) - shows the inset text (insert 2 of the same files into the repository and Jsoup parse the page in the controller IT test)
multiple duplicates- shows the inset text (insert 2 sets of duplicates into the repository and Jsoup parse the page in the controller IT test)

**Comment by LewisMitchell** [comment]
Understandably this is covered in the AT's but we may want to cover it before it gets to that stage (failing fast)

> I feel like there is a test or 2 missing in the controller integration test? Could we add some tests that shows the inset text? So like: 2 duplicates (of the same file) - shows the inset text (insert 2 of the same files into the repository and Jsoup parse the page in the controller IT test) multiple duplicates- shows the inset text (insert 2 sets of duplicates into the repository and Jsoup parse the page in the controller IT test)

**Comment by BackRowOla** [comment]
> Understandably this is covered in the AT's but we may want to cover it before it gets to that stage (failing fast)
> 
> > I feel like there is a test or 2 missing in the controller integration test? Could we add some tests that shows the inset text? So like: 2 duplicates (of the same file) - shows the inset text (insert 2 of the same files into the repository and Jsoup parse the page in the controller IT test) multiple duplicates- shows the inset text (insert 2 sets of duplicates into the repository and Jsoup parse the page in the controller IT test)

Makes sense to include the tests in the IT's, as I wouldn't have caught the inset text issue if I had not run the AT's

**Review by LewisMitchell — APPROVED** [review]

**Comment by LewisMitchell** [comment]
In UpscanServiceSpec replace add another case after line 129:
`UploadJourney("file3", UploadStatusEnum.DUPLICATE),`

In OtherReasonControllerSpec after line 387:
```
      "return OK and correct view - for duplicate uploads" in new Setup(AuthTestModels.successfulAuthResult) {
        when(mockUploadJourneyRepository.getUploadsForJourney(any())).thenReturn(Future.successful(Some(Seq(duplicateCallBackModel))))
        val result: Future[Result] = controller.onPageLoadForUploadComplete(NormalMode)(userRequestWithCorrectKeys)
        status(result) shouldBe OK
      }
```
In OtherReasonController, line 225 to be replaced with:
   ```
     if(previousUploadsState.isEmpty ||
          previousUploadsState.exists(!_.exists(file => file.fileStatus == UploadStatusEnum.READY || file.fileStatus == UploadStatusEnum.DUPLICATE))) {
```
In UpscanService, line 116 to be replaced with:
```      
_.fold(0)(_.count(file => file.fileStatus == UploadStatusEnum.READY || file.fileStatus == UploadStatusEnum.DUPLICATE))
```

**Review by LewisMitchell — CHANGES_REQUESTED** [review]
See latest comment

File: test/controllers/OtherReasonControllerSpec.scala [hunk 19]
```diff
@@ -324,307 +324,313 @@ class OtherReasonControllerSpec extends SpecBase {
...
-          status(result) shouldBe SEE_OTHER
-          redirectLocation(result).get shouldBe controllers.routes.OtherReasonController.onPageLoadForFirstFileUpload(NormalMode).url
+          status(result) shouldBe OK
         }
 
-        "there is no successful uploads" in new Setup(AuthTestModels.successfulAuthResult) {
-          val duplicateCallbackModel = UploadJourney("file1", UploadStatusEnum.FAILED, failureDetails = Some(FailureDetails(FailureReasonEnum.REJECTED, "upscan.invalidMimeType")))
-          when(mockUploadJourneyRepository.getUploadsForJourney(any())).thenReturn(Future.successful(Some(Seq(duplicateCallbackModel))))
+        "return OK and correct view - for duplicate uploads" in new Setup(AuthTestModels.successfulAuthResult) {
+          when(mockUploadJourneyRepository.getUploadsForJourney(any())).thenReturn(Future.successful(Some(Seq(callBackModel.copy(reference = "ref2")))))
```
Comment (LewisMitchell) [inline test/controllers/OtherReasonControllerSpec.scala:390]: Could it be 
```
Seq(
callBackModel.copy(reference = "ref2", fileStatus = UploadStatusEnum.DUPLICATE),
callBackModel.copy(reference = "ref3", fileStatus = UploadStatusEnum.DUPLICATE)
)
```
?

**Review by LewisMitchell — CHANGES_REQUESTED** [review]

**Review by LewisMitchell — APPROVED** [review]
